app/models/complaint.rb

```diff
@@ -0,0 +1,97 @@
+class Complaint
+  attr_reader :id, :attributes
+
+  EVENT_FIELDS = %w[
+    issueLastUpdated
+    creationDate
+    closedDate
+    reopenedDate
+    initialContactDate
+  ].freeze
+
+  EVENT_LABELS = {
+    issueLastUpdated: "Updated",
+    creationDate: "Created in HSES",
+    closedDate: "Closed",
+    reopenedDate: "Reopened",
+    initialContactDate: "Initial contact from complaint"
+  }.with_indifferent_access.freeze
+
+  def initialize(hses_complaint)
+    @id = hses_complaint[:id]
+    @attributes = hses_complaint[:attributes].with_indifferent_access
+  end
+
+  def events
+    @events ||= mapped_events
+      .sort_by { |events| events[:date] }
+      .reverse
+  end
+
+  def event_label(attribute_name)
+    EVENT_LABELS[attribute_name] || "Updated"
+  end
+
+  def grantee
+    attributes["grantee"]
+  end
+
+  def issue
+    attributes["issue"]
+  end
+
+  def due_date?
+    due_date.present?
+  end
+
+  def overdue?
+    due_date? && due_date.past?
+  end
+
+  def due_soon?
+    due_date? && due_date < 1.week.from_now
+  end
+
+  def relative_due_date_string
+    if due_date == Date.today
+      "Due today (#{formatted_due_date})"
+    else
+      "Due in #{day_string} (#{formatted_due_date})"
+    end
+  end
+
+  def formatted_due_date
+    due_date.strftime("%-m/%-d/%Y")
+  end
+
+  private
+
+  def relative_time_till_due
+    (Date.today...due_date).count
```

Comment:
this is only called if there is a due date

Reply:
ah i messed up the order in the template so that it currently can be called ... i will fix that along with other things to ensure this is not called if there is no due date.

we could still have a guard for it, but i don't think it'll be necessary